test/liquidity_gap.test.ts

```diff
@@ -0,0 +1,201 @@
+      true,
+      slippage
+    )
+
+    await initSwap(
+      invariant,
+      swapper,
+      tokenX,
+      tokenY,
+      fee,
+      tickSpacing,
+      true,
+      amount,
+      true,
+      targetSqrtPrice
+    )
+
+    const pool = await getPool(invariant, tokenX, tokenY, fee, tickSpacing)
+    const expectedSqrtPrice = (
+      await invariant.methods.calculateSqrtPrice({ args: { tickIndex: -10n } })
+    ).returns
+    const expectedYAmountOut = 9999n
+    const liquidityDelta = 20006000n * 10n ** LiquidityScale
+    const lowerTick = -10n
+
+    expect(pool.sqrtPrice).toEqual(expectedSqrtPrice)
+    expect(pool.liquidity).toEqual(liquidityDelta)
+    expect(pool.currentTickIndex).toEqual(lowerTick)
+
+    const swapperX = await balanceOf(tokenX.contractId, swapper.address)
+    const swapperY = await balanceOf(tokenY.contractId, swapper.address)
+    const dexX = await balanceOf(tokenX.contractId, invariant.address)
+    const dexY = await balanceOf(tokenY.contractId, invariant.address)
+
+    const deltaDexX = dexX - dexXBefore
+    const deltaDexY = dexYBefore - dexY
+
+    expect(swapperX).toEqual(0n)
+    expect(swapperY).toEqual(expectedYAmountOut + amount)
+    expect(deltaDexX).toEqual(amount)
+    expect(deltaDexY).toEqual(expectedYAmountOut)
+    expect(pool.feeGrowthGlobalX).toEqual(29991002699190242927121n)
+    expect(pool.feeGrowthGlobalY).toEqual(0n)
+    expect(pool.feeProtocolTokenX).toEqual(1n)
+    expect(pool.feeProtocolTokenY).toEqual(0n)
+  })
+  test('Open second position non-adjacent to the previous one, consequently creating a gap in liquidity', async () => {
+    const lowerTick = -90n
+    const upperTick = -50n
+    const liquidityDelta = 20008000n * 10n ** LiquidityScale
+    const amount = 10n ** 6n
+    await withdrawTokens(swapper, [tokenX, amount], [tokenY, amount])
+
+    const poolBefore = await getPool(invariant, tokenX, tokenY, fee, tickSpacing)
+
+    const [slippageLimitLower, slippageLimitUpper] = [poolBefore.sqrtPrice, poolBefore.sqrtPrice]
+
+    await initPositionWithLiquidity(
+      invariant,
+      swapper,
+      tokenX,
+      amount,
+      tokenY,
+      amount,
+      fee,
+      tickSpacing,
+      lowerTick,
+      upperTick,
+      liquidityDelta,
+      1n,
+      slippageLimitLower,
+      slippageLimitUpper
+    )
+  })
+  test('should skip gap and then swap', async () => {
+    const amount = 10067n
+    await withdrawTokens(swapper, [tokenX, amount], [tokenY, amount])
+
+    const slippage = MinSqrtPrice
+
+    const { targetSqrtPrice } = await quote(
+      invariant,
+      swapper,
+      tokenX,
+      tokenY,
+      fee,
+      tickSpacing,
+      true,
+      amount,
+      true,
+      slippage
+    )
+
+    await initSwap(
+      invariant,
+      swapper,
+      tokenX,
+      tokenY,
+      fee,
+      tickSpacing,
+      true,
+      amount,
+      true,
+      targetSqrtPrice
+    )
+
+    const pool = await getPool(invariant, tokenX, tokenY, fee, tickSpacing)
```

Comment:
It looks like you skipped a lot of important checks. Additionally, carefully check the area around the cross ticks to ensure that the three cases that may arise are properly implemented